nemo/core/classes/modelPT.py

```diff
@@ -157,86 +164,79 @@ def register_artifact(self, config_path: str, src: str):
+        # we are assuming that the location of the right nemo file is available from _MODEL_RESTORE_PATH
+        elif src.startswith("nemo:"):
+            tmpdir = _CACHE_ROOT + f"_TMPDIRS/{uuid.uuid4().hex}"
+            os.makedirs(tmpdir)
+            # TODO: Write a code to delete this in destructor
+            outfolder = self._unpack_nemo_file(path2file=_MODEL_RESTORE_PATH, out_folder=tmpdir)
+            #print(f" AAA: {os.listdir(tmpdir)}, B: {src}")
+            return_path = os.path.abspath(os.path.join(outfolder, src[5:]))
+        else:
+            raise FileNotFoundError(f"src path does not exist or it is not a path in nemo file. src value I got was: {src}")
+
+        assert os.path.exists(return_path)
+        artifact_item = model_utils.ArtifactItem()
+        artifact_item.path = os.path.abspath(src)
+        artifact_item.path_type = model_utils.ArtifactPathType.LOCAL_PATH
+        self.artifacts[config_path] = artifact_item
+        return return_path
+
+    def _handle_artifacts(self, nemo_file_folder):
+        for conf_path, artiitem in self.artifacts.items():
+            if artiitem.path_type == model_utils.ArtifactPathType.LOCAL_PATH:
+                if not os.path.exists(artiitem.path):
+                    raise FileNotFoundError(f"Artifact {conf_path} not found at location: {artiitem.path}")
+
+                # Generate new uniq artifact name and copy it to nemo_file_folder
+                # Note uuid.uuid4().hex is guaranteed to be 32 character long
+                artifact_base_name = os.path.basename(artiitem.path)
+                artifact_uniq_name = f"{uuid.uuid4().hex}_{artifact_base_name}"
+                shutil.copy2(artiitem.path, os.path.join(nemo_file_folder, artifact_uniq_name))
+
+                # Update artifacts registry
+                new_artiitem = model_utils.ArtifactItem()
+                new_artiitem.path = "nemo:" + artifact_uniq_name
+                new_artiitem.path_type = model_utils.ArtifactPathType.TAR_PATH
+                self.artifacts[conf_path] = new_artiitem
 
-            # But disregarding whether you use "local" or "remote" artifact - always store the original path.
-            # This fixes issues raising when finetuning NLP models that create and register tokenizer vocabs.
-            if config_path in self.artifacts:
-                logging.warning(
-                    f"Artifact {config_path} with value '{self.artifacts[config_path]}' "
-                    f"already exists and will be overwritten with value '{src}'!"
-                )
+            else:
+                raise ValueError(f"Directly referencing artifacts from other nemo files isn't supported yet")
 
-            self.artifacts[config_path] = archive_item
-            return used_src
-        else:
-            return src
+    def _update_artifact_paths(self, path2yaml_file):
+        if self.artifacts is not None and len(self.artifacts) > 0:
```

Comment:
This defers updating of config to the moment the file is written, which means at least some training time has already gone. I would rather overwrite self.cfg.* with the artifact paths at the moment the register method is called, so that the actual config is updated. Then the normal save will use the self.cfg to write the config.

Reply:
overwriting self.cfg would be wrong because model will be still in use. save_to hashes artifact names and replaces them with paths starting with "nemo:" specifically for the purposes of .nemo file

Reply:
Ok that makes sense